plasmapy/diagnostics/proton_radiography.py

```diff
@@ -0,0 +1,1022 @@
+        else:
+            gyroperiod = 2 * np.pi * self._m_p / (self._e * np.max(Bmag))
+
+        # TODO: introduce a minimum timestep based on electric fields too!
+
+        # Create an array of all the possible time steps we computed
+        candidates = np.array([gyroperiod / 12, gridstep])
+
+        # Enforce limits on dt
+        candidates = np.clip(candidates, self.dt[0], self.dt[1])
+
+        # dt is the min of the remaining candidates
+        return np.min(candidates)
+
+    def _coast_to_grid(self):
+        r"""
+        Coasts all particles to the timestep when the first particle should
+        be entering the grid. Doing in this in one step (rather than pushing
+        the particles through zero fields) saves computation time.
+        """
+        # Distance from the source to the nearest gridpoint
+        dist = np.min(np.linalg.norm(self.grid_arr - self.source, axis=3))
+
+        # Find the particle with the highest speed towards the grid
+        vmax = np.max(np.dot(self.v, self.src_n))
+
+        # Time for fastest possible particle to reach the grid.
+        t = dist / vmax
+
+        # Coast the particles to the advanced position
+        self.x = self.x + self.v * t
+
+    def _generate_null_distribution(self):
+        r"""
+        Calculate the distribution of particles on the detector plane in the absence
+        of any simulated fields.
+
+        These positions are used to quickly compute null radiographs, which are
+        used to determine the degree of deflection.
+        """
+        # Calculate the unit vector from the source to the detector
+        dist = np.linalg.norm(self.src_det)
+        uvec = self.src_det / dist
+
+        # Calculate the remaining distance each particle needs to travel
+        # along that unit vector
+        remaining = np.dot(self.source, uvec)
+
+        # Calculate the time remaining to reach that plane and push
+        t = (dist - remaining) / np.dot(self.v, uvec)
+
+        # Calculate the particle positions for that case
+        self.x0 = self.source + self.v * np.outer(t, np.ones(3))
+
+    def _coast_to_detector(self):
+        r"""
+        Coasts all particles to the detector plane. This method will be
+        called after all particles have cleared the grid.
+
+        This step applies to all particles, including those that never touched
+        the grid.
+        """
+        dist_remaining = np.dot(self.x, self.det_n) + np.linalg.norm(self.detector)
+
+        v_towards_det = np.dot(self.v, -self.det_n)
+
+        # Time remaining for each particle to reach detector plane
+        t = dist_remaining / v_towards_det
+
+        # If particles have not yet reached the detector plane and are moving
+        # away from it, they will never reach the detector.
+        # So, we can remove them from the arrays
+        condition = np.logical_and(v_towards_det < 0, dist_remaining > 0)
+        ind = np.nonzero(np.where(condition, 0, 1))[0]
```

Comment:
If we want to pick particles for which both conditions are true, this simplifies to...

```suggestion
        ind = (v_towards_det < 0) & (dist_remaining > 0)
```
And boolean indexing does the rest!

Reply:
Hm, for some reason I didn't think these logical operators worked element-wise on arrays...but I guess they do! 

The actual one-line replacement should be 
```
ind = np.logical_not((v_towards_det < 0) & (dist_remaining > 0)).nonzero()[0]
```
though, since `ind` should be the indices of all of the values where the statement is False.